core/src/test/java/com/linecorp/armeria/client/PendingAcquisitionTest.java

```diff
@@ -0,0 +1,98 @@
+
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+
+import com.linecorp.armeria.common.AggregatedHttpResponse;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.util.Exceptions;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelOutboundHandlerAdapter;
+import io.netty.channel.ChannelPromise;
+
+class PendingAcquisitionTest {
+
+    @RegisterExtension
+    static final ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) {
+            sb.http(0);
+            sb.https(0);
+            sb.tlsSelfSigned();
+            sb.service("/", (ctx, req) -> HttpResponse.of(200));
+        }
+    };
+
+    @CsvSource({ "H2C,HTTP", "H2,HTTPS" })
+    @ParameterizedTest
+    void shouldUsePendingAcquisitionFromExplicitProtocol(SessionProtocol explicitProtocol,
+                                                         SessionProtocol nonExplicitProtocol) {
+        final CountingConnectionPoolListener connectionPoolListener = new CountingConnectionPoolListener();
+
+        final PendingAcquisitionHandler pendingAcquisitionHandler = new PendingAcquisitionHandler();
+        ClientFactoryOptions.CHANNEL_PIPELINE_CUSTOMIZER.doNewValue(pipeline -> {
+            pipeline.addLast(pendingAcquisitionHandler);
+        });
+        try (ClientFactory factory = ClientFactory.builder()
+                                                  .connectionPoolListener(connectionPoolListener)
+                                                  .tlsNoVerify()
+                                                  .build()) {
+            final WebClient client = WebClient.builder()
+                                              .factory(factory)
+                                              .build();
+            final CompletableFuture<AggregatedHttpResponse> responseH2 =
+                    client.get(server.uri(explicitProtocol).resolve("/").toString())
+                          .aggregate();
+            final CompletableFuture<AggregatedHttpResponse> responseHttp =
+                    client.get(server.uri(nonExplicitProtocol).resolve("/").toString())
+                          .aggregate();
+            pendingAcquisitionHandler.latch.complete(null);
+            assertThat(responseH2.join().status()).isEqualTo(HttpStatus.OK);
+            assertThat(responseHttp.join().status()).isEqualTo(HttpStatus.OK);
+            assertThat(connectionPoolListener.opened()).isEqualTo(1);
+        }
+    }
+
+    static class PendingAcquisitionHandler extends ChannelOutboundHandlerAdapter {
```

Comment:
Is this used?

Reply:
It is used to defer connection acquisition until all requests are submitted.